### PR TITLE
Improve shader preprocessor with regex-fu

### LINKED ISSUE #3961: shaders preprocessor messup with GLSL texture functions (texture -> texMap String replace)
Eg. textureOffset() becomes texMapOffset()
I'm using a workaround (by letting preprocessor replace texMapCube to texture, so texMapCubeOffset() becomes textureOffset()

I think you could simply add the texture functions to Search and Replace Strings in the shaders pre-processor setup

### LINKED ISSUE #3968: GLSL pre-processing step changes names of variables named textureXX to texMapXX
**The shader doesn't have a uniform called "textureB" OR the uniform was removed during compilation because it was unused.**

It's used for sure!

Mac OS X, Processing 3

````java
PShader psBlendLightest;

PGraphics pg;
PGraphics textureB;


void settings() {
  size(512, 512, P2D);
}

void setup() {

  pg = createGraphics(512, 512, P2D);  
  pg.beginDraw();
  pg.background(0);
  pg.fill(255);
  pg.quad(25, 25, 400, 40, 450, 500, 40, 300);
  pg.endDraw();
  
  textureB = createGraphics(512, 512, P2D); 
  textureB.beginDraw();
  textureB.background(0);
  textureB.fill(255);
  textureB.ellipse(256, 256, 400, 400);
  textureB.endDraw();
  
  //psBlendLightest = loadShader("http://localhost:8888/shaders/blendLightest.glsl");
  psBlendLightest = new PShader(this, "http://localhost:8888/shaders/blendLightest.vert", "http://localhost:8888/shaders/blendLightest.frag");
  
}


void draw() {
  image(textureB, 0, 0);  
  psBlendLightest.set("textureB", textureB);
  filter(psBlendLightest);
}
````
blendLightest.frag
````
uniform sampler2D texture;
uniform sampler2D textureB;
 
varying vec4 vertTexCoord;
 
void main() {
    vec3 texColorA = texture2D(texture, vertTexCoord.xy).rgb;
    vec3 texColorB = texture2D(textureB, vertTexCoord.xy).rgb;
    gl_FragColor = vec4(max(texColorA, texColorB), 1.0);
}
````

blendLightest.vert
````
uniform mat4 transform;
 
attribute vec4 vertex;
attribute vec4 texCoord;
 
varying vec4 vertTexCoord;
 
void main() {
    vertTexCoord = vec4(vec2(texCoord.x, 1.0 - texCoord.y), 1.0, 1.0);
    gl_Position = transform * vertex;
}
````

### PR DESCRIPTION
**Pull request by JakubValtar**
Context is now taken into account, so it's not just blind search and replace anymore.

Fixes #3961
Fixes #3968

### CODE REVIEW

File: core/src/processing/opengl/PGL.java [hunk 1]
```diff
@@ -1883,23 +1885,20 @@ protected static int qualityToSamples(int quality) {
...
+          Pattern.compile(String.format(GLSL_ID_REGEX, "texture")),
+          Pattern.compile(String.format(GLSL_FN_REGEX, "textureRect|texture2D|texture3D|textureCube")),
+          Pattern.compile(String.format(GLSL_ID_REGEX, "gl_FragColor"))
       };
       String[] replace = new String[] {
-          "in", "in",
-          "texMap",
-          "texture", "texture", "texture", "texture",
-          "fragColor"
+          "in", "texMap", "texture", "_fragColor"
```
Comment (codeanticode) [inline core/src/processing/opengl/PGL.java]: @JakubValtar why do you use _fragColor instead of fragColor?

Comment (JakubValtar) [inline core/src/processing/opengl/PGL.java]: To make it more unique. Somebody could already have fragColor variable in their code and we don't have collision detection yet.

Comment (JakubValtar) [inline core/src/processing/opengl/PGL.java]: @codeanticode Doesn't matter what the name is really. It just has to be "out"

Comment (codeanticode) [inline core/src/processing/opengl/PGL.java]: Yes, I agree.

**Pull request merged**